File: java/L100/LeetCode0144.java
package L100;

import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

import util.TreeNode;

public class LeetCode0144 {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        if (root == null) {
            return res;
        }
        Deque<TreeNode> stack = new LinkedList<>();
        stack.add(root);
        while (!stack.isEmpty()) {
            TreeNode cur = stack.removeFirst();
            if (cur.right != null) {
                stack.addFirst(cur.right);
            }
            if (cur.left != null) {
                stack.addFirst(cur.left);
            }
            res.add(cur.val);
        }
        return res;

    }
}

File: cpp/L100/LeetCode0144.cc
#include "../heads/tree_node.h"
#include <stack>
#include <vector>
using std::vector;

class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> res;
        std::stack<TreeNode *> st;
        TreeNode *node = root;
        while (!st.empty() || node != nullptr) {
            if (node == nullptr) {
                node = st.top();
                st.pop();
            }
            res.push_back(node->val);
            if (node->right != nullptr) {
                st.push(node->right);
            }
            if (node->left != nullptr) {
                st.push(node->left);
            }
            node =  nullptr;
        }
        return res;
    }
};